File: leetcode/cpp/876. 链表的中间结点.cpp
typedef struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
} ListNode;

class Solution {
public:
    // 偶数时，返回后一个节点
    ListNode* middleNode(ListNode* head) {
        ListNode* p = head;
        ListNode* q = head;
        while (q && q->next) {
            p = p->next;
            q = q->next->next;
        }
        return p;
    }

    // 偶数时，返回前一个节点
    ListNode* middleNode2(ListNode* head) {
        ListNode* p = head;
        ListNode* q = head;
        while (q && q->next && q->next->next) {
            p = p->next;
            q = q->next->next;
        }
        return p;
    }
};

File: leetcode/python/876. 链表的中间结点.py
#!/bin/bash python
"""
876. 链表的中间结点

双指针法，快慢指针
"""

class ListNode(object):
    def __init__(self, x=0):
        self.val = x
        self.next = None

class Solution(object):
    def middleNode(self, head):
        """
        当偶数时，返回后一个节点
        """
        p = q = head
        while q and q.next:
            p = p.next
            q = q.next.next
        return p
    
    def middleNode2(self, head):
        """
        当偶数时，返回前一个节点
        """
        p = q = head
        while q and q.next and q.next.next:
            p = p.next
            q = q.next.next
        return p